File: app/src/main/java/com/reddittask/android/ui/activity/main/MainPresenterImpl.java
package com.reddittask.android.ui.activity.main;

import com.reddittask.android.api.Queries;
import com.reddittask.android.model.GameData;
import com.reddittask.android.ui.activity.base.MvpView;

/**
 * Created by marinaracu on 30.11.2017.
 */

public class MainPresenterImpl implements MainPresenter, Queries.CallBack {

    private MainView mMainView;
    private Queries mQueries;


    public MainPresenterImpl(MainView mainView, Queries queries, int dataCountLimit) {
        mMainView = mainView;
        mQueries = queries;
        mQueries.getAllDataMovies(dataCountLimit);
        mQueries.setCallBack(this);
    }


    @Override
    public void onAttach(MvpView mvpView) {
        mvpView.showLoading();
    }

    @Override
    public void onDetach() {
        mQueries.onStop();
    }

    @Override
    public void onItemNewsClicked(int position, String url) {
        mMainView.openDetailGame(position, url);
    }

    @Override
    public void loadNextData(int dataCountLimit) {
        mQueries.getAllDataMovies(dataCountLimit);
    }


    @Override
    public void onError(Throwable throwable) {
        mMainView.showErrorMessage();
        mMainView.hideLoading();
    }

    @Override
    public void onCompleted(GameData gameData) {
        mMainView.displayGameData(gameData);
    }

    @Override
    public void onFinish() {
        mMainView.hideLoading();
    }
}

File: app/src/main/java/com/keddits/ui/activity/main/MainPresenterImpl.kt
package com.keddits.ui.activity.main

import com.keddits.api.Queries
import com.keddits.model.Child
import com.keddits.model.GameData
import com.keddits.rx.HelperRx

/**
 * Created by marinaracu on 25.12.2017.
 */
class MainPresenterImpl(var queries: Queries, var dataCountLimit: Int) : MainPresenter<MainView>, Queries.CallBack, HelperRx.HelperRxCallBack {
    private lateinit var  mainView : MainView
    private lateinit var helperRx : HelperRx

    override fun onAttach(mvpView: MainView) {
        mainView = mvpView
        mainView.showLoadding()
        queries.getAllDatamovie(dataCountLimit)
        queries.setCallBack(this)
        helperRx = HelperRx()
        helperRx.setHelperRxCalback(this)
    }

    override fun onDetach() {
        queries.onStop()
    }

    override fun getView(): MainView {
        return mainView
    }

    override fun onItemNewsClicked(url: String) {
        mainView.openDetailGame(url)
    }

    override fun loadLastItems(childList: MutableList<Child>) {
        mainView.showLastItemLists(childList)
    }

    override fun loadNextData(dataCountLimit: Int, exitingItems: Int) {
        queries.getAllDatamovie(dataCountLimit)
    }

    override fun loadLastData(initialList: MutableList<Child>, skipCount: Int) {
        helperRx.getLastItemList(initialList, skipCount)
    }

    override fun onError(throwable: Throwable) {
        mainView.showErrorMessage()
        mainView.hideLoadding()
    }

    override fun onCompleted(gameData: GameData) {
        mainView.displayGameData(gameData.data.children)
    }

    override fun onFinish() {
        mainView.hideLoadding()
    }
}